src/vtpty/SshSession.cpp

```diff
@@ -0,0 +1,1168 @@
+                    _p->sshChannel,
+                    unbox<int>(_pageSize.columns),
+                    unbox<int>(_pageSize.lines),
+                    _pixels.has_value() ? unbox<int>(_pixels->width) : 0,
+                    _pixels.has_value() ? unbox<int>(_pixels->height) : 0);
+                if (rc == LIBSSH2_ERROR_EAGAIN)
+                {
+                    _p->wantsWaitForSocket = true;
+                    return;
+                }
+                if (rc != LIBSSH2_ERROR_NONE)
+                {
+                    logError("Failed to request PTY resize. {}", libssl2ErrorString(rc));
+                    return;
+                }
+                setState(State::Operational);
+                return;
+            }
+            case State::Failure:
+            case State::Closed:
+                //.
+                return;
+        }
+    }
+}
+
+void SshSession::start()
+{
+    if (_port == 22)
+        logInfo("Starting SSH session to host: {}@{}", _username, _host);
+    else
+        logInfo("Starting SSH session to host: {}@{}:{}", _username, _host, _port);
+
+    assert(_state == State::Initial);
+    // auto const _ = std::lock_guard { _mutex };
+    setState(State::Started);
+    processState();
+
+    /*
+        if (!_p->sshClient.connect(_host, _port))
+            return;
+
+        if (!_p->sshClient.handshake())
+            return;
+
+        if (!_p->sshClient.authenticate(_username))
+        {
+            sshLog()("SSH agent based authentication failed. Trying again with password.");
+            if (!_p->sshClient.authenticate(_username, "winki"))
+                return;
+        }
+
+        if (!_p->sshClient.openChannel())
+            return;
+
+        if (SshClient::supportsAuthAgentForwarding())
+            _p->sshClient.requestAuthAgentForwarding(); // Continue on failure
+
+        if (!_p->sshClient.requestPty(_env.count("TERM") ? _env.at("TERM") : "",
+                                   _pageSize.columns.as<int>(),
+                                   _pageSize.lines.as<int>(),
+                                   _pixels.has_value() ? _pixels->width.as<int>() : 0,
+                                   _pixels.has_value() ? _pixels->height.as<int>() : 0))
+            return;
+
+        _p->sshClient.setEnvironment(_env);
+
+        if (!_p->sshClient.startShell())
+            return;
+
+        // _p->sshClient.setBlocking(false);
+    */
+}
+
+PtySlave& SshSession::slave() noexcept
+{
+    return *_ptySlave; // Can we find a way to avoid this?
+}
+
+void SshSession::close()
+{
+    setState(State::Closed);
+
+    if (_p->sshChannel)
+    {
+        libssh2_channel_send_eof(_p->sshChannel);
+        libssh2_channel_close(_p->sshChannel);
+        libssh2_channel_wait_closed(_p->sshChannel);
+    }
+
+    if (_p->sshSocket.is_open())
+        _p->sshSocket.close();
+}
+
+void SshSession::waitForClosed()
+{
+    // TODO: find better implementation (using polling)
```

Comment:
```suggestion
    // TODO(pr): find better implementation (using polling)
```
or as a separate one?

Reply:
i just did it right away (using condition variable rather than polling)